ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -0,0 +1,500 @@
+        continue;
+      }
+      String replicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
+      if (!replicasStr.isEmpty()) {
+        for (String replicaInfo : replicasStr.split(ClusterMapUtils.REPLICAS_DELIM_STR)) {
+          String[] info = replicaInfo.split(ClusterMapUtils.REPLICAS_STR_SEPARATOR);
+          // partition name and replica name are the same.
+          String partitionName = info[0];
+          if (currentReplicasOnNode.containsKey(partitionName)) {
+            // if replica is already present
+            AmbryReplica existingReplica = currentReplicasOnNode.get(partitionName);
+            // 1. directly add it into "replicasFromConfig" map
+            replicasFromConfig.put(partitionName, existingReplica);
+            // 2. update replica seal/stop state
+            updateReplicaStateAndOverrideIfNeeded(existingReplica, sealedReplicas, stoppedReplicas);
+          } else {
+            // if this is a new replica and doesn't exist on node
+            logger.info("Adding new replica {} to existing node {}", partitionName, instanceName);
+            long replicaCapacity = Long.valueOf(info[1]);
+            String partitionClass = info.length > 2 ? info[2] : clusterMapConfig.clusterMapDefaultPartitionClass;
+            // this can be a brand new partition that is added to an existing node
+            AmbryPartition mappedPartition =
+                new AmbryPartition(Long.valueOf(partitionName), partitionClass, helixClusterManagerCallback);
+            // Ensure only one AmbryPartition instance exists for specific partition.
+            mappedPartition = clusterChangeHandlerCallback.addPartitionIfAbsent(mappedPartition, replicaCapacity);
+            ensurePartitionAbsenceOnNodeAndValidateCapacity(mappedPartition, dataNode, replicaCapacity);
+            // create new replica belonging to this partition
+            AmbryReplica replica =
+                new AmbryReplica(clusterMapConfig, mappedPartition, disk, stoppedReplicas.contains(partitionName),
+                    replicaCapacity, sealedReplicas.contains(partitionName));
+            updateReplicaStateAndOverrideIfNeeded(replica, sealedReplicas, stoppedReplicas);
+            // add new created replica to "replicasFromConfig" map
+            replicasFromConfig.put(partitionName, replica);
+            // add new replica to specific partition
+            clusterChangeHandlerCallback.addReplicasToPartition(mappedPartition, Collections.singletonList(replica));
+          }
+        }
+      }
+    }
+    // update ambryDataNodeToAmbryReplicas map by adding "replicasFromConfig"
+    ambryDataNodeToAmbryReplicas.put(instanceNameToAmbryDataNode.get(instanceName), replicasFromConfig);
+    // compare replicasFromConfig with current replica set to derive old replicas that are removed
+    Set<AmbryReplica> previousReplicas = new HashSet<>(currentReplicasOnNode.values());
+    previousReplicas.removeAll(replicasFromConfig.values());
```

Comment:
could this be done with partition IDs without the copy? 
```java
currentReplicasOnNode.keySet()
  .stream()
  .filter(partitionId -> !replicasFromConfig.containsKey(partitionId))
  .forEach(do delete);
```

Reply:
Changed